skymp5-server/cpp/lib_espm/impl/MappedBuffer.cpp

```diff
@@ -0,0 +1,99 @@
+#include "MappedBuffer.h"
+
+#include <fmt/format.h>
+
+#ifndef WIN32
+#  include <iostream>
+
+// Linux
+#  include <fcntl.h>
+#  include <sys/mman.h>
+#  include <unistd.h>
+#endif
+
+namespace espm::impl {
+
+MappedBuffer::MappedBuffer(const fs::path& path)
+{
+  size_ = fs::file_size(path);
+#ifdef WIN32
+  fileHandle_ = CreateFileW(path.c_str(), GENERIC_READ, FILE_SHARE_READ, NULL,
+                            OPEN_EXISTING, FILE_FLAG_RANDOM_ACCESS, 0);
+  if (!fileHandle_) {
+    // generic vs system?
+    throw std::system_error(
+      std::error_code(GetLastError(), std::system_category()),
+      "CreateFileW failed");
+  }
+
+  mapHandle_ = CreateFileMapping(fileHandle_, NULL, PAGE_READONLY, 0, 0, NULL);
+  if (!mapHandle_) {
+    throw std::system_error(
+      std::error_code(GetLastError(), std::system_category()),
+      "CreateFileMapping failed");
+  }
+
+  viewPtr_ = MapViewOfFile(mapHandle_, FILE_MAP_READ, 0, 0, 0);
+  if (!mapHandle_) {
+    throw std::system_error(
+      std::error_code(GetLastError(), std::system_category()),
+      "CreateFileMapping failed");
+  }
+  data_ = static_cast<char*>(viewPtr_);
+#else
+  fd_ = open(path.c_str(), O_RDONLY);
+  if (fd_ == -1) {
+    throw std::system_error(errno, std::generic_category(),
+                            fmt::format("Can't read {}", path.string()));
+  }
+  const auto mmapResult = mmap(NULL, size_, PROT_READ, MAP_SHARED, fd_, 0);
+  if (mmapResult == MAP_FAILED) {
+    throw std::system_error(errno, std::generic_category(),
+                            fmt::format("Can't map {}", path.string()));
+  }
+  data_ = static_cast<char*>(mmapResult);
+#endif
+}
+
+MappedBuffer::~MappedBuffer()
+{
+#ifdef WIN32
+  if (viewPtr_) {
+    UnmapViewOfFile(viewPtr_);
```

Comment:
`UnmapViewOfFile` and `CloseHandle` may also fail.

Reply:
UPD: Let's think `CloseHandle` may not fail. No one in the real world checks it.

Reply:
But actually it can fail in theory (I guess only in case when invalid handle passed, but still).

So, I think, we can check it just in case (and calling `std::abort` if any of these fail, because it seems like an extremely abnormal situation).

Reply:
A small note.

Both of these functions' return values are unchecked in this answer: https://stackoverflow.com/a/13059914

They use `if (FHandle != traits::InvalidValue)` check instead. I guess it's the only reason this function should fail.

As for Linux's `close` syscall, it can fail in some circumstances, such as wrong fd, interrupted syscall or I/O error (e.g. on network drives): https://stackoverflow.com/questions/19056309/not-checking-closes-return-value-how-serious-really

As for `munmap`, it could also fail if incorrect params passed or in some cases with intersecting regions https://stackoverflow.com/questions/43743555/munmap-failure-with-enomem-with-private-anonymous-mapping

I think these cases should not come up for us, but it's better if we'll know somehow about it